resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java

```diff
@@ -829,50 +968,175 @@ void add(DnsRecord r) {
+                        serverName.next = server;
+                        serverName = server;
+
+                        servers++;
+                    }
+                    // We should replace the TTL if needed with the one of the ADDITIONAL record so we use
+                    // the smallest for caching.
+                    serverName.update(parent.newRedirectServerAddress(resolved), r.timeToLive());
+                    return;
                 }
                 serverName = serverName.next;
             }
-            return null;
         }
 
-        int size() {
-            return count;
+        // Now handle all AuthoritativeNameServer for which we had no ADDITIONAL record
+        void handleWithoutAdditionals(DnsNameResolver parent, DnsCache cache) {
+            AuthoritativeNameServer serverName = head;
+
+            while (serverName != null) {
+                if (serverName.address == null) {
+                    // Try to resolve via cache
+
+                    List<? extends DnsCacheEntry> entries = cache.get(serverName.nsName, null);
+                    if (entries == null || entries.isEmpty()) {
+                        // These will be resolved on the fly if needed.
+                        serverName.address = InetSocketAddress.createUnresolved(
+                                serverName.nsName, DefaultDnsServerAddressStreamProvider.DNS_PORT);
+                        servers++;
+                    } else {
+                        // We have some entries in the cache.
+                        for (int i = 0; i < entries.size(); i++) {
+                            InetAddress address = entries.get(i).address();
+                            if (address != null) {
+                                if (serverName.address != null) {
+                                    AuthoritativeNameServer server = new AuthoritativeNameServer(serverName);
+                                    server.next = serverName.next;
+                                    serverName.next = server;
+                                    serverName = server;
+
+                                    servers++;
+                                }
+                                // Use a TTL of Long.MIN_VALUE as we don't want to cache the actual address later on as
+                                // we have no idea about how long the TTL should be. When we encounter Long.MIN_VALUE
+                                // later on we will cache the unresolved address.
+                                serverName.update(parent.newRedirectServerAddress(address), Long.MIN_VALUE);
+
+                                ++i;
+                                for (; i < entries.size(); i++) {
+                                    address = entries.get(i).address();
+                                    // If the first address was not null the rest should not be null as well but
+                                    // let us just check.
```

Comment:
+1 for the comment ... would this be a candidate for an `assert` or do you think it is legit?